Dockerfile

```diff
@@ -21,72 +24,29 @@ SHELL ["/bin/bash", "-o", "pipefail", "-c"]
 # --- Install Basic Software --------------------
 # -----------------------------------------------
 
-RUN \
-  apt-get -qq update && \
-  apt-get -qq install apt-utils 2>/dev/null && \
-  apt-get -qq dist-upgrade && \
-  echo "applying workaround for ubuntu/postfix bug described in https://github.com/docker-mailserver/docker-mailserver/issues/2023#issuecomment-855326403" && \
-  mv /bin/hostname{,.bak} && \
-  echo "echo docker-mailserver.invalid" > /bin/hostname && \
-  chmod +x /bin/hostname && \
-  apt-get -qq install postfix && \
-  mv /bin/hostname{.bak,} && \
-  apt-get -qq --no-install-recommends install \
-  # A - D
-  altermime amavisd-new apt-transport-https arj binutils bzip2 bsd-mailx \
-  ca-certificates cabextract clamav clamav-daemon cpio curl \
-  dbconfig-no-thanks dovecot-core dovecot-fts-xapian dovecot-imapd \
-  dovecot-ldap dovecot-lmtpd dovecot-managesieved dovecot-pop3d \
-  dovecot-sieve dovecot-solr dumb-init \
-  # E - O
-  ed fetchmail file gamin gnupg gzip iproute2 \
-  locales logwatch lhasa libdate-manip-perl libldap-common liblz4-tool \
-  libmail-spf-perl libnet-dns-perl libsasl2-modules lrzip lzop \
-  netcat-openbsd nftables nomarch opendkim opendkim-tools opendmarc \
-  # P - Z
-  pax pflogsumm postgrey p7zip-full postfix-ldap postfix-pcre \
-  postfix-policyd-spf-python postsrsd pyzor \
-  razor rpm2cpio rsyslog sasl2-bin spamassassin supervisor \
-  unrar-free unzip uuid whois xz-utils && \
-  # Fail2Ban
-  gpg --keyserver ${FAIL2BAN_GPG_PUBLIC_KEY_SERVER} \
-    --recv-keys ${FAIL2BAN_GPG_PUBLIC_KEY_ID} 2>&1 && \
-  curl -Lkso fail2ban.deb ${FAIL2BAN_DEB_URL} && \
-  curl -Lkso fail2ban.deb.asc ${FAIL2BAN_DEB_ASC_URL} && \
-  FINGERPRINT=$(LANG=C gpg --verify \
-  fail2ban.deb.asc fail2ban.deb 2>&1 \
-    | sed -n 's#Primary key fingerprint: \(.*\)#\1#p') && \
-  if [[ -z ${FINGERPRINT} ]]; then \
-    echo "ERROR: Invalid GPG signature!" >&2; exit 1; fi && \
-  if [[ ${FINGERPRINT} != "${FAIL2BAN_GPG_FINGERPRINT}" ]]; then \
-    echo "ERROR: Wrong GPG fingerprint!" >&2; exit 1; fi && \
-  dpkg -i fail2ban.deb 2>&1 && \
-  rm fail2ban.deb fail2ban.deb.asc && \
-  # cleanup
-  apt-get -qq autoremove && \
-  apt-get -qq autoclean && \
-  apt-get -qq clean && \
-  rm -rf /var/lib/apt/lists/* && \
-  c_rehash 2>&1
-
+COPY target/scripts/build/* /build/
 COPY target/scripts/helpers/log.sh /usr/local/bin/helpers/log.sh
-COPY target/bin/sedfile /usr/local/bin/sedfile
+RUN /bin/bash /build/packages.sh
```

Comment:
Have you looked into if this invalidates cache when a change is made in `packages.sh` for a new build?

This is not an approach I recall seeing elsewhere when that matters. And Dockers ["Best Practices" guide for build cache](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#leverage-build-cache) states:

> Aside from the `ADD` and `COPY` commands, cache checking does not look at the files in the container to determine a cache match.
> For example, when processing a `RUN apt-get -y update` command the files updated in the container are not examined to determine if a cache hit exists.
> In that case just **the command string itself is used to find a match**.

I think this isn't an issue though since we're using an earlier `COPY` for the script, which will perform an invalidation AFAIK.

---

Due to the dependence on `log.sh`, any changes to that is going to invalidate this layer as well which is a shame, but hopefully rarely an issue going forward.

Reply:
There is [one way around the issue with cache invalidation](https://docs.docker.com/engine/reference/builder/#run---mounttypecache), which local builds could at least take advantage of. I don't think it'd affect our CI caching negatively either.

Just for reference, as it's mostly only going to matter for local builds if you're frequently adjusting `packages.sh` or `log.sh`. The linked docs provide an example specifically with `apt` usage:

```Dockerfile
# syntax=docker/dockerfile:1
FROM ubuntu
RUN rm -f /etc/apt/apt.conf.d/docker-clean; echo 'Binary::apt::APT::Keep-Downloaded-Packages "true";' > /etc/apt/apt.conf.d/keep-cache
RUN --mount=type=cache,target=/var/cache/apt,sharing=locked \
  --mount=type=cache,target=/var/lib/apt,sharing=locked \
  apt update && apt-get --no-install-recommends install -y gcc
```

> _`RUN --mount` allows you to create mounts that process running as part of the build can access._
> _This can be used to bind files from other part of the build without copying, accessing build secrets or ssh-agent sockets, or creating cache locations to speed up your build._
> 
> _`RUN --mount=type=cache`_
> _This mount type allows the build container to cache directories for compilers and package managers._

No need to look into that for this PR, just making a note of it here.

That'd change how we're handling post-install cleanup I think, so isn't appropriate for this PR where we want to keep large diffs of shuffling lines around easy to track before further revision.

I'm not sure where it mounts that locally, presumably a docker data / volume container instead of a bind volume mount.

Reply:
The `COPY` will indeed invalidate cache, so the `RUN` is executed again. And `log.sh` is rarely touched, so this is a non-issue if you ask me.

---

👍🏼 when it comes to the solution about the APT package cache. Might do another PR in the future for that.